Always reject Promises with Error instances

felixfbecker opened the pull request:
I noticed that in many places `Promise.wrapError` is called with a string, which means the Promise will be rejected with a string instead of an `Error` instance. One might think that `wrapError` wraps the string in an Error, but it really is just a `Promise.reject()`. Rejecting with a string is like `throw`ing a string, it means no call stack, no metadata like `severity`, `code` or `actions` and possible bugs when assuming the value is an Error when it's not (e.g. `err.message` being `undefined`).

These were all the places I found with a regexp search (doesn't catch if the string is saved in a variable). I did not find any `throw` statements with a string.

jrieken commented:
Not yet convinced by the arguments but surely it doesn't make sense to have a new programming-rule without a corresponding tslint-rule. These changes will erode quickly

felixfbecker commented:
If you throw strings like
```ts
throw 'Something went wrong'
```
wherever that error bubbles up or is caught, it will have no `.stack` callstack. This makes it very hard to debug.

When an error is caught with `catch` you know can no longer assume the message is always under `err.message`, because if it is a string, it will be the message itself, meaning `typeof` checks are needed.

`Error` objects being objects also allow you to attach more metadata. VS Code uses `err.severity` and `error.actions` to determine how to show the error to the user. Now, if you want to add one of these later on to a thrown Error, that is no problem. But if you throw a string, you now need to refactor a whole bunch of code because you can no longer use a string and need to change all the `catch` blocks that relied on this or did `typeof` checks.

On the other hand, always being consistent has no disadvantage to it.

Now, my example was for `throw`, but `Promise.wrapError()` is equivalent to throwing/catching like above in an `async`/`await` function, they are the same concepts.

TSLint has a rule for the string `throw`: https://palantir.github.io/tslint/rules/no-string-throw/

> Flags throwing plain strings or concatenations of strings because only Errors produce proper stack traces.

But `Promise.wrapError()` is harder to catch, not sure how to do that?

jrieken commented:
> But Promise.wrapError() is harder to catch, not sure how to do that?

We have a set of hand-written tslint rules for special cases. A new rule must be implemented in `build/lib/tslint`. Before starting on this a team decision has to be made. I'll bring it up for discussion during our standup

jrieken commented:
Ok, proposal is disallow throwing something which isn't an error and to disallow rejecting a promise with some that isn't an error. To enforce these rules the following rules/changes are required

* have a lint rule that enforces proper usage of the `throw`-statement
* have a lint rule that does the same for `TPromise.wrapError` *or* modify the `winjs.base.d.ts` file which spec's this. Something like this `public static wrapError<V, E extends Error>(error: E): TPromise<V>;` and `	constructor(init: (complete: TValueCallback<V>, error: <T>(err: <T extends Error>) => void, progress: ProgressCallback) => void, oncancel?: any);
`

The Zürich-team expresses like for this, /cc @kieferrm for the collecting the Redmond take on this. The question remains how far we take this. Correct would be to have not a `TPromise<T>` but something like `TPromise<V, E extends Error>` and `wrapError<E extends Error>(error: E): Promise<never, E>`

felixfbecker commented:
Great to hear!
I would not type the rejection type, because it breaks the symmetry to sync functions. In opposite to Java, where Exception types are declared with `throws` and part of the interface, they are not (yet?) in TypeScript: https://github.com/Microsoft/TypeScript/issues/13219.
So if you call into any external code somewhere, the type system doesn't know whether that function _might_ throw something other than an `Error`. Thrown exceptions are `any` in TS from a type perspective, it's only a bad practice to throw anything else in own code (and therefor suitable for a lint rule). For that reason all async equivalents in the wild have their Errors typed `any` as well, like `Promise` (from lib.d.ts), `Observable` etc.
So I wouldn't introduce a new generic argument for `Promise`, but I think adding a type constraint to `wrapError` is the easiest way forward.

felixfbecker commented:
@jrieken Could you tell me what meaning the boolean `silent` rejection value here has? https://sourcegraph.com/github.com/Microsoft/vscode/-/blob/src/vs/workbench/api/node/extHostTextEditor.ts#L544-545

felixfbecker commented:
`onError` here is typed as returning `TPromise<IConfigurationEditingError>`, but only ever rejects the Promise, never fulfills it with that type. Also unsure why it returns a Promise as it's not async.  https://sourcegraph.com/github.com/Microsoft/vscode/-/blob/src/vs/workbench/services/configuration/node/configurationEditingService.ts#L109-122

felixfbecker commented:
I would like to use `Object.assign()`, but TypeScript complains that it isn't available. As Electron is using Node 7, can I add `lib: es2015` to tsconfig.json?

joaomoreno commented:
cc @sandy081 For the last comment.

jrieken commented:
re https://github.com/Microsoft/vscode/pull/29034#issuecomment-309823655 - The idea is to fail silent or not. Tho since we are always silent, so don't really have a use-case, I'd suggest to nuke it

sandy081 commented:
re [comment](https://github.com/Microsoft/vscode/pull/29034#issuecomment-309861209) `onError` method does not need to return anything. Idea is to whether to notify error, so this can return `void`.

jrieken commented on src/vs/base/common/winjs.base.d.ts:
```diff
@@ -134,5 +134,5 @@ export declare class PPromise<C, P> extends TPromise<C> {
 	public static join<C, P>(promises: { [n: string]: PPromise<C, P> }): PPromise<{ [n: string]: C }, P>;
 	public static any<C, P>(promises: PPromise<C, P>[]): PPromise<{ key: string; value: PPromise<C, P>; }, P>;
-	public static wrapError<V>(error: any): TPromise<V>;
+	public static wrapError(error: Error): TPromise<never>;
```
Removing the generic type is the cause for all those compile errors. While it seems correct, I'd go the way [`lib.d.ts`](https://github.com/Microsoft/TypeScript/blob/master/lib/lib.es2015.promise.d.ts#L207) did and have both signatures. Especially, when returning a rejected-promise from a function that returns `Promise<T>` isn't possible with the `never`-approch.

felixfbecker replied:
I know, I have uncommitted changes locally :)

>  Especially, when returning a rejected-promise from a function that returns Promise<T> isn't possible with the never-approch.

Could you clarify what you mean by this? One of the traits of the `never` type is that you can always assign/return it, e.g. a function can have a return type `Promise<number>` but you can return `Promise<never>`. Just like a function can return `number`, but actually throw an Error in some cases (which means its real return type is `number | never`, which is equivalent to `number`, because `never` always disappears in union types).

lib.d.ts actually will return `Promise<never>` if you don't pass a type argument: https://sourcegraph.com/github.com/Microsoft/TypeScript@master/-/blob/lib/lib.es2015.promise.d.ts?q=reject#L200-201
I am pretty sure they only left the old version for backwards-compatibility as the `never` type is quite new

jrieken replied:
Actually true, wasn't aware that `never` is assignable everywhere. This might be our first usage of never ...

jrieken commented on tslint.json:
```diff
@@ -3,6 +3,7 @@
 		"build/lib/tslint"
 	],
 	"rules": {
+		"no-string-throw": true,
```
Is that it? No smarter lint rules that actually check for a type?

felixfbecker replied:
The rule will catch all kinds of string literals and concatenation of strings. It will probably not trigger if you save the string to a variable first. It sure could be extended to do that, but I am not familiar enough with custom TSLint rules to do that

https://sourcegraph.com/github.com/palantir/tslint@master/-/blob/src/rules/noStringThrowRule.ts#L61:11

felixfbecker replied:
I went manually through all occurrences of `throw` in the codebase and found 2 cases. Not sure about this one:
https://sourcegraph.com/github.com/Microsoft/vscode/-/blob/extensions/typescript/src/features/implementationsCodeLensProvider.ts#L45-46
@mattbierner

mjbvz replied:
This one should probably not even use a `throw`. In this PR, you can either leave it as is or convert it to something like `throw new Error('Could not resolve codelens')`

felixfbecker commented:
All tests pass.
- Removed the `silent` parameter https://github.com/Microsoft/vscode/pull/29034#issuecomment-309986452
- Made `onError` return `void` https://github.com/Microsoft/vscode/pull/29034#issuecomment-310080475
- For the cases where objects were thrown, I replaced the interface with an Error subclass with the same interface.

jrieken commented:
Great, please resolve merge conflicts and will make sure to get this quickly

jrieken commented on src/vs/base/common/winjs.base.d.ts:
```diff
@@ -106,7 +106,7 @@ export declare class TPromise<V> {
 	public static wrap<ValueType>(value: ValueType): TPromise<ValueType>;
 
-	public static wrapError<ValueType>(error: any): TPromise<ValueType>;
+	public static wrapError<ValueType>(error: Error): TPromise<ValueType>;
```
what happened to `wrapError(error: Error): TPromise<never>`?

felixfbecker replied:
I met some compile errors I couldn't resolve, I think due to `Thenable` / `TPromise` / `PPromise` all being different types. If it's okay I would like to keep this PR at only the dont-reject-with-strings change.

felixfbecker commented:
Rebased and fixed merge conflicts, build is failing because master is failing.

jrieken commented:
Sorry, one more rebase/resolve conflicts is needed, I'll try merge quickly then to make sure this catchup game eventually stops

felixfbecker commented:
Done. Build failure seems to be cause master is broken

jrieken commented:
Thanks. Merging now
